### PR TITLE
[SPARK-20214][ML] Make sure converted csc matrix has sorted indices

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

`_convert_to_vector` converts a scipy sparse matrix to csc matrix for initializing `SparseVector`. However, it doesn't guarantee the converted csc matrix has sorted indices and so a failure happens when you do something like that:

    from scipy.sparse import lil_matrix
    lil = lil_matrix((4, 1))
    lil[1, 0] = 1
    lil[3, 0] = 2
    _convert_to_vector(lil.todok())

    File "/home/jenkins/workspace/python/pyspark/mllib/linalg/__init__.py", line 78, in _convert_to_vector
      return SparseVector(l.shape[0], csc.indices, csc.data)
    File "/home/jenkins/workspace/python/pyspark/mllib/linalg/__init__.py", line 556, in __init__
      % (self.indices[i], self.indices[i + 1]))
    TypeError: Indices 3 and 1 are not strictly increasing

A simple test can confirm that `dok_matrix.tocsc()` won't guarantee sorted indices:

    >>> from scipy.sparse import lil_matrix
    >>> lil = lil_matrix((4, 1))
    >>> lil[1, 0] = 1
    >>> lil[3, 0] = 2
    >>> dok = lil.todok()
    >>> csc = dok.tocsc()
    >>> csc.has_sorted_indices
    0
    >>> csc.indices
    array([3, 1], dtype=int32)

I checked the source codes of scipy. The only way to guarantee it is `csc_matrix.tocsr()` and `csr_matrix.tocsc()`.

## How was this patch tested?

Existing tests.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by viirya** [comment]
cc @jkbradley @MLnick @holdenk

**Comment by srowen** [comment]
I don't think is directly related, but, I tried to make a similar change to the Spark Scala implementation of this type of conversion and couldn't make it work, just because it changed a bunch of little behaviors. (I could look up the JIRA.) This could be fine, being more narrow.

File: python/pyspark/ml/linalg/__init__.py [hunk 1]
```diff
@@ -72,7 +72,9 @@ def _convert_to_vector(l):
         return DenseVector(l)
     elif _have_scipy and scipy.sparse.issparse(l):
         assert l.shape[1] == 1, "Expected column vector"
-        csc = l.tocsc()
+        # Make sure the converted csc_matrix has sorted indices.
+        csc = l.tocsr().tocsc()
```
Comment (jkbradley) [inline python/pyspark/ml/linalg/__init__.py]: I'd just call tocsc.  You can then check the attribute has_sorted_indices and call sort_indices() if needed.

**Comment by jkbradley** [comment]
Are you able to write a unit test which passes data through _convert_to_vector and fails before this fix?

**Comment by jkbradley** [comment]
Btw, I'd really like to get this into 2.2, which will be cut soon.  Let me know if you'd like me to take it over.  Thanks!

**Comment by viirya** [comment]
@jkbradley Thanks for comment. I will add the unit test now.

File: python/pyspark/ml/linalg/__init__.py (hunk 1 above)
Comment (viirya) [inline python/pyspark/ml/linalg/__init__.py]: Thanks. Fixed.

**Comment by viirya** [comment]
@jkbradley An unit test is added. Please check this again. Thanks.

File: python/pyspark/mllib/tests.py [hunk 2]
```diff
@@ -853,6 +853,17 @@ def serialize(l):
         self.assertEqual(sv, serialize(lil.tocsr()))
         self.assertEqual(sv, serialize(lil.todok()))
 
+    def test_convert_to_vector(self):
```
Comment (viirya) [inline python/pyspark/mllib/tests.py]: We have no Scipy-related tests in `ml` submodule, so I don't add it. If we really need one, please let me know. I can add it quickly.

Comment (jkbradley) [inline python/pyspark/mllib/tests.py]: No it's Ok; I made a separate JIRA to port the tests there.

**Comment by jkbradley** [comment]
LGTM
Merging with master and branch-2.1, branch-2.0
Thanks a lot!

**Pull request closed without merging**